### DISCUSSION TITLE
Release of Docker Container

### DISCUSSION THREAD

**Discussion by peterdavidfagan** [isaac-sim/IsaacLab#14] [Ideas]
Will there be a Docker container released for Isaac Sim with Orbit included or should users expect to manually build upon the currently available container for Isaac Sim? I presume the latter but wished to check in case there was already existing work in this direction.

Thanks in advance for any clarity you can provide.

**Comment by ltorabi** [comment]
You are right, we do not have a container with Orbit included yet.

**Reply by peterdavidfagan** [reply]
Thanks for the quick response @ltorabi. 

I am starting to build a Dockerfile for this purpose. I probably need to check the license agreement for the Nvidia Isaac Sim image that is being pulled but I'd happy to open a pr with the Dockerfile all going well.

**Reply by sheikh-nv** [reply]
See this repo for an example to build your own container.
https://github.com/NVIDIA-Omniverse/IsaacSim-dockerfiles

**Reply by animesh-garg** [reply]
@peterdavidfagan that would be a grewat contrib.
indeed please do.

**Reply by peterdavidfagan** [reply]
I posted https://github.com/NVIDIA-Omniverse/Orbit/pull/17 for the above. I don't believe it is ready for review yet, I wish to play with some of the orbit scripts and make sure that everything is working as expected. Also I haven't completed the checklist items.

@sabdulajees I built directly on top of your image and reference it from https://github.com/NVIDIA-Omniverse/IsaacSim-dockerfiles as I think this is the cleanest way to build the orbit container from the file you shared. Let me know if you think there is a better way to do this.

Update: I have an initial container config created that works on core isaac sim demos. Unfortunately I have encountered some issues when running the Orbit demos with this config, I opened issues on this repository to communicate what I have encountered. I'll be returning to this later in the week as having a containerised environment for Orbit would be useful for my use case.

**Comment by Mayankm96** [comment]
Hi @peterdavidfagan ,

Thanks a lot for taking this initiative :)

Sorry for the late response. I was away on vacation. Let me know how I can help in finalizing the docker container for orbit in the coming week.

**Reply by peterdavidfagan** [reply]
Thanks @Mayankm96 and I hope you had a great vacation. 

Admittedly I haven't been as active with this docker container development recently; instead I have started to work towards creating/contributing our own environments for Orbit (using a workstation installation of orbit).

The status when I last left this was that there were some errors in pulling assets when building upon ``Dockerfile.2022.1.1-ubuntu20.04``. The same issues aren't present for 2022.2 however an official Dockerfile isn't present for this version. Naively updating the version on the existing [Dockerfile](https://github.com/NVIDIA-Omniverse/IsaacSim-dockerfiles/blob/main/Dockerfile.2022.1.1-ubuntu20.04) didn't build for me last time I checked but I haven't invested much time in debugging this. @sabdulajees mentioned that there would be an official Dockerfile for 2022.2 in the near future, I think once this is available updating the current pr to build upon this should be sufficient. 

If there are any design flaws you wish to be fixed in the current pr I'd be happy to address these.

**Reply by AndrejOrsula** [reply]
Hello,

This Dockerfile works for me using `nvcr.io/nvidia/isaac-sim:2022.2.0` as the base (parent) image:
```Dockerfile
ARG ISAACSIM_VERSION=2022.2.0
FROM nvcr.io/nvidia/isaac-sim:${ISAACSIM_VERSION}

### Use bash by default
SHELL ["/bin/bash", "-c"]

### Install Orbit
RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive apt-get install -yq --no-install-recommends \
        cmake \
        ncurses-term \
        git && \
    rm -rf /var/lib/apt/lists/* && \
    git clone https://github.com/NVIDIA-Omniverse/Orbit.git --depth 1 -b main /orbit && \
    ln -s "/isaac-sim" /orbit/_isaac_sim && \
    sed -i "s/pip install pre-commit/$(extract_isaacsim_python) -m pip install pre-commit/" /orbit/orbit.sh && \
    /orbit/orbit.sh --install && \
    /orbit/orbit.sh --extra || exit 0

### Set entrypoint (adjust to your preferences)
ENTRYPOINT ["/orbit/orbit.sh", "-p"]
```

There are a few peculiarities though:

#### 1) Because `orbit.sh --install` tries to install `pre-commit` using system-wide `pip` installation/alias, this step fails while building the Docker image because the parent image does not ship with system-wide `pip`.

https://github.com/NVIDIA-Omniverse/Orbit/blob/66d5b603e5958c1b3c3ebc2a7bbb66011c2b10eb/orbit.sh#L114

Instead of installing `python3-pip` via `apt` like in the current [PR](https://github.com/NVIDIA-Omniverse/Orbit/pull/17/files#diff-5206fae30d1f89c1fd2424c5dccb3db75c260d6dd1538176848c01892094fd44R36), I substitute the `pip install pre-commit` in the installation script with `$(extract_isaacsim_python) -m pip install pre-commit`.

```Dockerfile
    sed -i "s/pip install pre-commit/$(extract_isaacsim_python) -m pip install pre-commit/" /orbit/orbit.sh && \
```

However, **I would suggest placing the installation of `pre-commit`** either **inside the existing `-f|--format` case**, another flag or a separate script altogether, as it is not really essential for end-users of Orbit (only contributors).

<details>
  <summary>(click to expand) Personally, I place the installation of pre-commit under a separate `.git_hooks/install.bash` script that also installs the git hooks</summary>

```bash
set -e

SCRIPT_DIR="$(cd "$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")" &>/dev/null && pwd)"
REPO_DIR="$(dirname "${SCRIPT_DIR}")"

# Install pre-commit if needed
if ! command -v pre-commit &>/dev/null; then
    pip install --user pre-commit
fi

# Install local git hooks for this repository
cd "${REPO_DIR}"
pre-commit install --install-hooks --config "${REPO_DIR}/.pre-commit-config.yaml"
cd - >/dev/null
```

... but I am not sure what is the standard approach.
</details>

#### 2) Installation of some of the extra modules during `orbit.sh --extra` failed for me while building the Docker image

I can't remember which ones, but I do not utilise (all) of them anyway. So I just suppressed the errors (not an ideal solution)...

```Dockerfile
    /orbit/orbit.sh --extra || exit 0
```

**Reply by Mayankm96** [reply]
Thanks @AndrejOrsula!

I agree this pre-commit is mostly a convenience for developers. It should be hidden cleanly and I like your suggestion to have a `-f/--format` flag. 

Would you have the bandwidth to send a PR with these changes?

**Reply by AndrejOrsula** [reply]
Sure, https://github.com/NVIDIA-Omniverse/Orbit/pull/28 is now open for this change.

**Comment by neale** [comment]
Related to the availability of a docker image. I also found that building off the isaac-sim image and installing Orbit works just fine. 

Are there plans to support livestreaming to OmniverseStreamingClient? Its offered in the [gym envs repo](https://github.com/NVIDIA-Omniverse/OmniIsaacGymEnvs). It would be really useful for the (I assume) many people who develop on remote servers.

**Reply by Mayankm96** [reply]
That's great to hear. We will take a look at updating the ongoing PR for the docker early next week and update the docs.

Regarding livestream, I have been preparing some things for that in #63. It should be straightforward to enable livestream. You can check #70 to keep track of its status.

**Reply by neale** [reply]
Thanks! I had previously tried something like:
```python
from omni.isaac.core.utils.extensions import enable_extension
enable_extension("omni.kit.livestream.native")
```
But I had errors resembling `extension 'omni.kit.livestream.native' not found`

I should become more familiar with `carb`

**Reply by Mayankm96** [reply]
Let's discuss this on the issue I made. Carb kind of goes towards the intricasies of how the simulator is handled. I don't think you need to dig that deep to make this work.

**Comment by Mayankm96** [comment]
We have now added docker support. Please check here for more details: https://isaac-orbit.github.io/orbit/source/deployment/index.html
